File: frontend/src/api/scheduled-generator.ts
/**
 * 定时自动生成器 API
 */
import { API_BASE_URL, API_PREFIX } from './novel'

const SCHEDULED_GEN_BASE = `${API_BASE_URL}${API_PREFIX}/scheduled-generator`

// ========== 类型定义 ==========

export interface ScheduledGeneratorConfig {
  id: number
  user_id: number
  name: string
  enabled: boolean
  trigger_hour: number
  trigger_minute: number
  generation_mode: 'basic' | 'enhanced'
  auto_upload_fanqie: boolean
  upload_interval_seconds: number
  concurrent_books: number
  max_duration_per_book: number
  status: 'idle' | 'running' | 'paused'
  total_runs: number
  total_books_processed: number
  total_books_success: number
  total_books_failed: number
  last_run_at: string | null
  next_run_at: string | null
  created_at: string
  updated_at: string
}

export interface QueueItem {
  id: number
  config_id: number
  novel_id: string
  priority: number
  position: number
  status: 'pending' | 'running' | 'completed' | 'failed' | 'skipped'
  auto_generator_task_id: number | null
  started_at: string | null
  completed_at: string | null
  duration_seconds: number | null
  chapters_generated: number
  chapters_target: number
  uploaded_to_fanqie: boolean
  upload_result: string | null
  error_message: string | null
  retry_count: number
  created_at: string
  updated_at: string
}

export interface SchedulerStatus {
  config_id: number
  status: string
  enabled: boolean
  next_run_at: string | null
  last_run_at: string | null
  queue_stats: {
    pending: number
    running: number
    completed: number
    failed: number
    total: number
  }
  currently_running: Array<{
    novel_id: string
    started_at: string | null
    chapters_generated: number
    chapters_target: number
  }>
  total_stats: {
    total_runs: number
    total_books_processed: number
    total_books_success: number
    total_books_failed: number
  }
}

export interface ExecutionLog {
  id: number
  config_id: number
  trigger_type: string
  trigger_time: string
  status: string
  books_processed: number
  books_success: number
  books_failed: number
  books_skipped: number
  total_chapters: number
  total_duration_seconds: number
  details: string | null
  error_message: string | null
  completed_at: string | null
  created_at: string
}

export interface ConfigCreateRequest {
  name: string
  trigger_hour: number
  trigger_minute: number
  generation_mode?: 'basic' | 'enhanced'
  auto_upload_fanqie?: boolean
  upload_interval_seconds?: number
  concurrent_books?: number
  max_duration_per_book?: number
}

export interface ConfigUpdateRequest {
  name?: string
  trigger_hour?: number
  trigger_minute?: number
  generation_mode?: 'basic' | 'enhanced'
  auto_upload_fanqie?: boolean
  upload_interval_seconds?: number
  concurrent_books?: number
  max_duration_per_book?: number
}

export interface AddToQueueRequest {
  novel_ids: string[]
  priority?: number
}

// ========== API 函数 ==========

const request = async (url: string, options: RequestInit = {}) => {
  const token = localStorage.getItem('token')
  const headers = new Headers({
    'Content-Type': 'application/json',
    ...options.headers
  })

  if (token) {
    headers.set('Authorization', `Bearer ${token}`)
  }

  const response = await fetch(url, { ...options, headers })

  if (response.status === 401) {
    throw new Error('未授权，请重新登录')
  }

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}))
    throw new Error(errorData.detail || `请求失败，状态码: ${response.status}`)
  }

  if (response.status === 204) {
    return null
  }

  return response.json()
}

export const scheduledGeneratorApi = {
  // ========== 配置管理 ==========

  async createConfig(data: ConfigCreateRequest): Promise<ScheduledGeneratorConfig> {
    return request(`${SCHEDULED_GEN_BASE}/config`, {
      method: 'POST',
      body: JSON.stringify(data)
    })
  },

  async getConfig(): Promise<ScheduledGeneratorConfig> {
    return request(`${SCHEDULED_GEN_BASE}/config`)
  },

  async updateConfig(configId: number, data: ConfigUpdateRequest): Promise<ScheduledGeneratorConfig> {
    return request(`${SCHEDULED_GEN_BASE}/config/${configId}`, {
      method: 'PUT',
      body: JSON.stringify(data)
    })
  },

  async deleteConfig(configId: number): Promise<void> {
    return request(`${SCHEDULED_GEN_BASE}/config/${configId}`, {
      method: 'DELETE'
    })
  },

  // ========== 队列管理 ==========

  async addToQueue(data: AddToQueueRequest): Promise<QueueItem[]> {
    return request(`${SCHEDULED_GEN_BASE}/queue/add`, {
      method: 'POST',
      body: JSON.stringify(data)
    })
  },

  async getQueue(): Promise<QueueItem[]> {
    return request(`${SCHEDULED_GEN_BASE}/queue`)
  },

  async removeFromQueue(itemId: number): Promise<void> {
    return request(`${SCHEDULED_GEN_BASE}/queue/${itemId}`, {
      method: 'DELETE'
    })
  },

  // ========== 任务控制 ==========

  async start(): Promise<{ status: string; message: string }> {
    return request(`${SCHEDULED_GEN_BASE}/start`, {
      method: 'POST'
    })
  },

  async pause(): Promise<{ status: string; message: string }> {
    return request(`${SCHEDULED_GEN_BASE}/pause`, {
      method: 'POST'
    })
  },

  async resume(): Promise<{ status: string; message: string }> {
    return request(`${SCHEDULED_GEN_BASE}/resume`, {
      method: 'POST'
    })
  },

  async stop(): Promise<{ status: string; message: string }> {
    return request(`${SCHEDULED_GEN_BASE}/stop`, {
      method: 'POST'
    })
  },

  async triggerNow(): Promise<{ status: string; message: string }> {
    return request(`${SCHEDULED_GEN_BASE}/trigger-now`, {
      method: 'POST'
    })
  },

  // ========== 状态查询 ==========

  async getStatus(): Promise<SchedulerStatus> {
    return request(`${SCHEDULED_GEN_BASE}/status`)
  },

  async getLogs(limit: number = 20): Promise<ExecutionLog[]> {
    return request(`${SCHEDULED_GEN_BASE}/logs?limit=${limit}`)
  }
}

File: backend/app/schemas/scheduled_generator.py
"""
定时生成器 Pydantic Schemas
"""
from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field, field_validator


class ScheduledGeneratorConfigCreate(BaseModel):
    """创建定时配置请求"""
    name: str = Field(..., description="配置名称")
    trigger_hour: int = Field(..., ge=0, le=23, description="触发小时（太平洋时间，0-23）")
    trigger_minute: int = Field(..., ge=0, le=59, description="触发分钟（0-59）")
    generation_mode: str = Field(default="basic", description="生成模式：basic/enhanced")
    auto_upload_fanqie: bool = Field(default=False, description="是否自动上传番茄小说")
    upload_interval_seconds: int = Field(default=20, ge=5, le=60, description="上传间隔（秒）")
    concurrent_books: int = Field(default=1, ge=1, le=5, description="并发生成书籍数（1-5）")
    max_duration_per_book: int = Field(default=120, ge=10, le=600, description="单本最大时长（分钟）")

    @field_validator("generation_mode")
    @classmethod
    def validate_generation_mode(cls, v: str) -> str:
        if v not in ["basic", "enhanced"]:
            raise ValueError("generation_mode must be 'basic' or 'enhanced'")
        return v


class ScheduledGeneratorConfigUpdate(BaseModel):
    """更新定时配置请求"""
    name: Optional[str] = None
    trigger_hour: Optional[int] = Field(None, ge=0, le=23)
    trigger_minute: Optional[int] = Field(None, ge=0, le=59)
    generation_mode: Optional[str] = None
    auto_upload_fanqie: Optional[bool] = None
    upload_interval_seconds: Optional[int] = Field(None, ge=5, le=60)
    concurrent_books: Optional[int] = Field(None, ge=1, le=5)
    max_duration_per_book: Optional[int] = Field(None, ge=10, le=600)

    @field_validator("generation_mode")
    @classmethod
    def validate_generation_mode(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and v not in ["basic", "enhanced"]:
            raise ValueError("generation_mode must be 'basic' or 'enhanced'")
        return v


class ScheduledGeneratorConfigResponse(BaseModel):
    """定时配置响应"""
    id: int
    user_id: int
    name: str
    enabled: bool
    trigger_hour: int
    trigger_minute: int
    generation_mode: str
    auto_upload_fanqie: bool
    upload_interval_seconds: int
    concurrent_books: int
    max_duration_per_book: int
    status: str
    total_runs: int
    total_books_processed: int
    total_books_success: int
    total_books_failed: int
    last_run_at: Optional[datetime] = None
    next_run_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class AddToQueueRequest(BaseModel):
    """添加到队列请求"""
    novel_ids: List[str] = Field(..., description="小说项目ID列表")
    priority: int = Field(default=0, description="优先级")


class QueueItemResponse(BaseModel):
    """队列项响应"""
    id: int
    config_id: int
    novel_id: str
    priority: int
    position: int
    status: str
    auto_generator_task_id: Optional[int] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    duration_seconds: Optional[int] = None
    chapters_generated: int
    chapters_target: int
    uploaded_to_fanqie: bool
    upload_result: Optional[str] = None
    error_message: Optional[str] = None
    retry_count: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class SchedulerStatusResponse(BaseModel):
    """调度器状态响应"""
    config_id: int
    status: str
    enabled: bool
    next_run_at: Optional[str] = None
    last_run_at: Optional[str] = None
    queue_stats: Dict[str, int]
    currently_running: List[Dict[str, Any]]
    total_stats: Dict[str, int]


class LogResponse(BaseModel):
    """日志响应"""
    id: int
    config_id: int
    trigger_type: str
    trigger_time: datetime
    status: str
    books_processed: int
    books_success: int
    books_failed: int
    books_skipped: int
    total_chapters: int
    total_duration_seconds: int
    details: Optional[str] = None
    error_message: Optional[str] = None
    completed_at: Optional[datetime] = None
    created_at: datetime

    model_config = {"from_attributes": True}
